_posts/2019-01-23-multifile-python-functions.md

```diff
@@ -0,0 +1,213 @@
+---
+title: Multifile python functions
+description: "Tidy python in an OpenFaaS world; or, what happens when you want to split your Python function into multiple modules?"
+date: 2019-01-23
+image: /images/multifile-python-function/pexels-book-stack-books-bookshop-264635.jpg
+categories:
+  - python
+author_staff_member: lucas
+dark_background: true
+---
+
+As a core contributor to OpenFaaS, I hangout in the OpenFaaS Slack a lot, you can [join us here][openfaas-slack-singnup]. Most recently that meant trying to answer this question
+
+> [How can I] write a function in python that uses multiple source files so that the main `handler.py` is able to import helper functions from separate source files.
+
+Today, I will walk though creating exactly this kind of function by creating a word count function.
+
+**TL;DR**: Use explicit relative imports to split your python code into several files. If you want to quickly jump to the finished code, I have create a [public repo here][project-repo].
+
+I assume some familiarity with OpenFaaS and that you may have even started/completed the [OpenFaaS Workshop][workshop-repo], so I won't go into a lot of detail about the specific OpenFaaS setup instructions below. Instead, I want to focus on how to comfortably develop a Python function as it goes from a small "one-liner" function to a larger multi-module project.
+
+## Set up a python environment
+
+This project uses Python 3 and I like to use Conda for my development environments:
+
+```sh
+$ conda create --name faaswordcloud  python=3.7.2 black autopep8 flake8 pylint mypy flask gevent
+$ conda source activate faaswordcloud
+```
+
+Or from my example repo: `conda env create -f environment.yml`.
+
+Creating a local environment like this will allow your editor/ide to lint/hint/autocomplete as needed.
+
+## Start the function
+
+Next, to start a new Python 3 function, I like to use the flask template because I am going to load a static list of "stop words" into memory when the function starts. Using the flask templates allows the function to do this only once instead of on each invocation.
+
+```sh
+$ faas-cli template pull https://github.com/openfaas-incubator/python-flask-template
+Fetch templates from repository: https://github.com/openfaas-incubator/python-flask-template at master
+2019/01/19 10:55:34 Attempting to expand templates from https://github.com/openfaas-incubator/python-flask-template
+2019/01/19 10:55:35 Fetched 3 template(s) : [python27-flask python3-flask python3-flask-armhf] from https://github.com/openfaas-incubator/python-flask-template
+$ faas-cli new wordcloud --lang python3-flask
+$ mv wordcloud.yml stack.yml
```

Comment:
Explain why

Reply:
Please make this edit to comment on why it's done, or don't do it. In my last post on secrets I didn't do it for example.